# Section 1: Dataframe of the postal codes, boroughs, and neighbourhoods of Toronto

In [1]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')

Solving environment: done

## Package Plan ##

  environment location: /home/jupyterlab/conda/envs/python

  added / updated specs:
    - geopy


The following packages will be UPDATED:

  ca-certificates     pkgs/main::ca-certificates-2020.1.1-0 --> conda-forge::ca-certificates-2020.4.5.2-hecda079_0
  certifi              pkgs/main::certifi-2020.4.5.1-py36_0 --> conda-forge::certifi-2020.4.5.2-py36h9f0ad1d_0

The following packages will be SUPERSEDED by a higher-priority channel:

  openssl              pkgs/main::openssl-1.1.1g-h7b6447c_0 --> conda-forge::openssl-1.1.1g-h516909a_0


Preparing transaction: done
Verifying transaction: done
Executing transaction: done
Solving environment: done

# All requested packages already installed.

Libraries imported.


In [2]:
import sys
!{sys.executable} --version
!pip install lxml

Python 3.6.10


In [3]:
df=pd.read_html("https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M")[0]
df.head()

Postal Code           Borough               Neighborhood
0         M1A      Not assigned               Not assigned
1         M2A      Not assigned               Not assigned
2         M3A        North York                  Parkwoods
3         M4A        North York           Victoria Village
4         M5A  Downtown Toronto  Regent Park, Harbourfront

In [4]:
df_del = df.drop(df[df['Borough']=='Not assigned'].index)
df_del.head()

Postal Code           Borough                                 Neighborhood
2         M3A        North York                                    Parkwoods
3         M4A        North York                             Victoria Village
4         M5A  Downtown Toronto                    Regent Park, Harbourfront
5         M6A        North York             Lawrence Manor, Lawrence Heights
6         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

The dataframe df_del has dropped all Boroughs with the value of Not assigned

In [20]:
df_reset = df_del.reset_index()
df_reset.drop("index", axis = 1, inplace=True)
df_reset.head()

Postal Code           Borough                                 Neighborhood
0         M3A        North York                                    Parkwoods
1         M4A        North York                             Victoria Village
2         M5A  Downtown Toronto                    Regent Park, Harbourfront
3         M6A        North York             Lawrence Manor, Lawrence Heights
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

df_reset has the index reset

In [13]:
df_reset.loc[df_reset['Neighborhood'] == 'Not assigned']

Empty DataFrame
Columns: [Postal Code, Borough, Neighborhood]
Index: []

Displaying rows that have a Neighborhood value of "Not assigned" does not exist

In [24]:
df_reset.groupby(['Postal Code']) #grouping by postal code so there are no duplicate postal codes
df_reset.head()

Postal Code           Borough                                 Neighborhood
0         M3A        North York                                    Parkwoods
1         M4A        North York                             Victoria Village
2         M5A  Downtown Toronto                    Regent Park, Harbourfront
3         M6A        North York             Lawrence Manor, Lawrence Heights
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

In [25]:
df_reset.shape

(103, 3)

**There are 103 rows (postal codes) in the dataframe.** 

# Section 2: Latitude and Longitude of the neighborhoods of Toronto

In [40]:
df_location= pd.read_csv('Geospatial_Coordinates.csv')
df_location.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [43]:
df_merge = pd.merge(df_location, df_reset, on=["Postal Code"])
df_merge.shape

(103, 5)

In [45]:
df_merge.head()

Postal Code   Latitude  Longitude      Borough  \
0         M1B  43.806686 -79.194353  Scarborough   
1         M1C  43.784535 -79.160497  Scarborough   
2         M1E  43.763573 -79.188711  Scarborough   
3         M1G  43.770992 -79.216917  Scarborough   
4         M1H  43.773136 -79.239476  Scarborough   

                             Neighborhood  
0                          Malvern, Rouge  
1  Rouge Hill, Port Union, Highland Creek  
2       Guildwood, Morningside, West Hill  
3                                  Woburn  
4                               Cedarbrae

# Section 3: Clustering of Neighborhoods and Visualizing on a Map

In [51]:
toronto_data = df_merge[df_merge['Borough'].str.contains("Toronto")].reset_index(drop=True)
toronto_data['Borough'].unique()

array(['East Toronto', 'Central Toronto', 'Downtown Toronto',
       'West Toronto'], dtype=object)

*Only using boroughs with the word "Toronto" in it*

In [52]:
toronto_data.shape

(39, 5)

In [54]:
address = 'Toronto, ON'

geolocator = Nominatim(user_agent="to_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


In [56]:
# create map of Toronto using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, label in zip(toronto_data['Latitude'], toronto_data['Longitude'], toronto_data['Neighborhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

*Map of the different neighborhoods of the borough of Toronto*

In [57]:
toronto_data.head()

Postal Code   Latitude  Longitude          Borough  \
0         M4E  43.676357 -79.293031     East Toronto   
1         M4K  43.679557 -79.352188     East Toronto   
2         M4L  43.668999 -79.315572     East Toronto   
3         M4M  43.659526 -79.340923     East Toronto   
4         M4N  43.728020 -79.388790  Central Toronto   

                     Neighborhood  
0                     The Beaches  
1    The Danforth West, Riverdale  
2  India Bazaar, The Beaches West  
3                 Studio District  
4                   Lawrence Park

In [112]:
# @hidden_cell
CLIENT_ID = 'C1DJSVQPIID5XJCYEBF3UJC2USPHCLQDDFCWPTSO45TB4EKE' # your Foursquare ID
CLIENT_SECRET = 'TZ5I5AOLSV2YH2APIIOE0NZG4NHBPOB5R1YRKTFNVHNTVC4P' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version
radius = 500
LIMIT = 50

If I have done this properly, the above cell with my client ID and client secret should be hidden, however I have used a radius of 500 meters and a LIMIT of 50

In [60]:
#function to get nearby venues for each neighborhood in toronto_data
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [64]:
#running the function above
toronto_venues = getNearbyVenues(names=toronto_data['Neighborhood'],
                                   latitudes=toronto_data['Latitude'],
                                   longitudes=toronto_data['Longitude']
                                  )

The Beaches
The Danforth West, Riverdale
India Bazaar, The Beaches West
Studio District
Lawrence Park
Davisville North
North Toronto West, Lawrence Park
Davisville
Moore Park, Summerhill East
Summerhill West, Rathnelly, South Hill, Forest Hill SE, Deer Park
Rosedale
St. James Town, Cabbagetown
Church and Wellesley
Regent Park, Harbourfront
Garden District, Ryerson
St. James Town
Berczy Park
Central Bay Street
Richmond, Adelaide, King
Harbourfront East, Union Station, Toronto Islands
Toronto Dominion Centre, Design Exchange
Commerce Court, Victoria Hotel
Roselawn
Forest Hill North & West, Forest Hill Road Park
The Annex, North Midtown, Yorkville
University of Toronto, Harbord
Kensington Market, Chinatown, Grange Park
CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport
Stn A PO Boxes
First Canadian Place, Underground city
Christie
Dufferin, Dovercourt Village
Little Portugal, Trinity
Brockton, Parkdale Village, Exhibition Place
High 

In [65]:
print(toronto_venues.shape)
toronto_venues.head()

(1191, 7)


Neighborhood  Neighborhood Latitude  \
0                   The Beaches              43.676357   
1                   The Beaches              43.676357   
2                   The Beaches              43.676357   
3                   The Beaches              43.676357   
4  The Danforth West, Riverdale              43.679557   

   Neighborhood Longitude                               Venue  Venue Latitude  \
0              -79.293031                   Glen Manor Ravine       43.676821   
1              -79.293031  The Big Carrot Natural Food Market       43.678879   
2              -79.293031                 Grover Pub and Grub       43.679181   
3              -79.293031                       Upper Beaches       43.680563   
4              -79.352188                       MenEssentials       43.677820   

   Venue Longitude     Venue Category  
0       -79.293942              Trail  
1       -79.297734  Health Food Store  
2       -79.297215                Pub  
3       -79.292869       Neighborhood  
4       -79.351265     Cosmetics Shop

*A table showing the nearby venues and their category for each neighborhood*

In [79]:
#analyze the venues in each neighborhood
# one hot encoding
tor_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

#dropping existing onehot neighborhood
tor_onehot_proper = tor_onehot.drop('Neighborhood', 1)

# add neighborhood column back to dataframe
tor_onehot_proper.insert(0, 'Neighborhood', toronto_venues['Neighborhood'])

tor_onehot_proper.head()

Neighborhood  Airport  Airport Food Court  Airport Gate  \
0                   The Beaches        0                   0             0   
1                   The Beaches        0                   0             0   
2                   The Beaches        0                   0             0   
3                   The Beaches        0                   0             0   
4  The Danforth West, Riverdale        0                   0             0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Antique Shop  Aquarium  Art Gallery  Art Museum  Arts & Crafts Store  \
0             0         0            0           0                    0   
1             0         0            0           0                    0   
2             0         0            0           0                    0   
3             0         0            0           0                    0   
4             0         0            0           0                    0   

   Asian Restaurant  Athletics & Sports  Auto Workshop  BBQ Joint  Baby Store  \
0                 0                   0              0          0           0   
1                 0                   0              0          0           0   
2                 0                   0              0          0           0   
3                 0                   0              0          0           0   
4                 0                   0              0          0           0   

   Bagel Shop  Bakery  Bank  Bar  Baseball Stadium  Basketball Stadium  Beach  \
0           0       0     0    0                 0                   0      0   
1           0       0     0    0                 0                   0      0   
2           0       0     0    0                 0                   0      0   
3           0       0     0    0                 0                   0      0   
4           0       0     0    0                 0                   0      0   

   Beer Bar  Beer Store  Belgian Restaurant  Bistro  Boat or Ferry  Bookstore  \
0         0           0                   0       0              0          0   
1         0           0                   0       0              0          0   
2         0           0                   0       0              0          0   
3         0           0                   0       0              0          0   
4         0           0                   0       0              0          0   

   Boutique  Brazilian Restaurant  Breakfast Spot  Brewery  Bubble Tea Shop  \
0         0                     0               0        0                0   
1         0                     0               0        0                0   
2         0                     0               0        0                0   
3         0                     0               0        0                0   
4         0                     0               0        0                0   

   Burger Joint  Burrito Place  Bus Line  Butcher  Café  \
0             0              0         0        0     0   
1             0              0         0        0     0   
2             0              0         0        0     0   
3             0              0         0        0     0   
4             0              0         0        0     0   

   Cajun / Creole Restaurant  Camera Store  Candy Store  Caribbean Restaurant  \
0                          0             0            0                     0   
1                          0             0            0                     0   
2                          0             0            0                     0   
3                    

In [80]:
#grouping venues by neighborhood and frequency
tor_grouped = tor_onehot_proper.groupby('Neighborhood').mean().reset_index()
tor_grouped.head()

Neighborhood   Airport  \
0                                        Berczy Park  0.000000   
1       Brockton, Parkdale Village, Exhibition Place  0.000000   
2  Business reply mail Processing Centre, South C...  0.000000   
3  CN Tower, King and Spadina, Railway Lands, Har...  0.055556   
4                                 Central Bay Street  0.000000   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0            0.000000      0.000000        0.000000         0.000000   
1            0.000000      0.000000        0.000000         0.000000   
2            0.000000      0.000000        0.000000         0.000000   
3            0.055556      0.055556        0.111111         0.166667   
4            0.000000      0.000000        0.000000         0.000000   

   Airport Terminal  American Restaurant  Antique Shop  Aquarium  Art Gallery  \
0          0.000000                  0.0           0.0       0.0         0.02   
1          0.000000                  0.0           0.0       0.0         0.00   
2          0.000000                  0.0           0.0       0.0         0.00   
3          0.111111                  0.0           0.0       0.0         0.00   
4          0.000000                  0.0           0.0       0.0         0.00   

   Art Museum  Arts & Crafts Store  Asian Restaurant  Athletics & Sports  \
0        0.00                  0.0               0.0                 0.0   
1        0.00                  0.0               0.0                 0.0   
2        0.00                  0.0               0.0                 0.0   
3        0.00                  0.0               0.0                 0.0   
4        0.02                  0.0               0.0                 0.0   

   Auto Workshop  BBQ Joint  Baby Store  Bagel Shop    Bakery  Bank       Bar  \
0       0.000000        0.0         0.0        0.02  0.040000   0.0  0.000000   
1       0.000000        0.0         0.0        0.00  0.083333   0.0  0.041667   
2       0.055556        0.0         0.0        0.00  0.000000   0.0  0.000000   
3       0.000000        0.0         0.0        0.00  0.000000   0.0  0.055556   
4       0.000000        0.0         0.0        0.00  0.000000   0.0  0.020000   

   Baseball Stadium  Basketball Stadium  Beach  Beer Bar  Beer Store  \
0               0.0                0.02   0.02      0.04         0.0   
1               0.0                0.00   0.00      0.00         0.0   
2               0.0                0.00   0.00      0.00         0.0   
3               0.0                0.00   0.00      0.00         0.0   
4               0.0                0.00   0.00      0.00         0.0   

   Belgian Restaurant  Bistro  Boat or Ferry  Bookstore  Boutique  \
0                 0.0    0.02       0.000000        0.0  0.000000   
1                 0.0    0.00       0.000000        0.0  0.000000   
2                 0.0    0.00       0.000000        0.0  0.000000   
3                 0.0    0.00       0.055556        0.0  0.055556   
4                 0.0    0.00       0.000000        0.0  0.000000   

   Brazilian Restaurant  Breakfast Spot   Brewery  Bubble Tea Shop  \
0                   0.0        0.020000  0.000000             0.00   
1                   0.0        0.083333  0.000000             0.00   
2                   0.0        0.000000  0.055556             0.00   
3                   0.0        0.000000  0.000000             0.00   
4                   0.0        0.000000  0.000000             0.04   

   Burger Joint  Burrito Place  Bus Line  Butcher   Café  \
0          0.00       0.000000       0.0      0.0  0.040   
1          0.00       0.041667       0.0      0.0  0.125   
2          0.00       0.055556       0.0      0.0  0.000   
3          0.00       0.000000       0.0      0.0  0.000   
4          0.04       0.000000       0.0      0.0  0.060   

   Cajun / Creole Restaurant  Camera Store  Candy Store  Caribbean Restaurant  \
0                        0.0           0.0          0.0             

In [81]:
#show top 5 type of venue in each neighborhood
num_top_venues = 5

for hood in tor_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = tor_grouped[tor_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Berczy Park----
                venue  freq
0         Coffee Shop  0.08
1        Cocktail Bar  0.06
2              Bakery  0.04
3  Seafood Restaurant  0.04
4         Cheese Shop  0.04


----Brockton, Parkdale Village, Exhibition Place----
            venue  freq
0            Café  0.12
1  Breakfast Spot  0.08
2          Bakery  0.08
3     Coffee Shop  0.08
4     Yoga Studio  0.04


----Business reply mail Processing Centre, South Central Letter Processing Plant Toronto----
                venue  freq
0  Light Rail Station  0.11
1         Yoga Studio  0.06
2              Garden  0.06
3         Pizza Place  0.06
4          Comic Shop  0.06


----CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport----
              venue  freq
0   Airport Service  0.17
1    Airport Lounge  0.11
2  Airport Terminal  0.11
3           Airport  0.06
4               Bar  0.06


----Central Bay Street----
                venue  freq
0         Coffee Sho

*The frequency of each of the top 5 venue categories in each neighborhood*

In [82]:
#put the data into dataframe
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [85]:
#show top 10 venues in each neighborhood in a dataframe
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = tor_grouped['Neighborhood']

for ind in np.arange(tor_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(tor_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                        Berczy Park           Coffee Shop   
1       Brockton, Parkdale Village, Exhibition Place                  Café   
2  Business reply mail Processing Centre, South C...    Light Rail Station   
3  CN Tower, King and Spadina, Railway Lands, Har...       Airport Service   
4                                 Central Bay Street           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0          Cocktail Bar           Cheese Shop                Bakery   
1                Bakery        Breakfast Spot           Coffee Shop   
2           Yoga Studio  Gym / Fitness Center  Fast Food Restaurant   
3        Airport Lounge      Airport Terminal               Airport   
4                  Café    Italian Restaurant        Sandwich Place   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0              Beer Bar    Seafood Restaurant                  Café   
1           Yoga Studio               Stadium         Burrito Place   
2        Farmers Market         Auto Workshop         Burrito Place   
3       Harbor / Marina           Coffee Shop                 Plane   
4       Bubble Tea Shop          Burger Joint                 Diner   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0            Restaurant        Clothing Store                  Diner  
1            Restaurant          Climbing Gym              Pet Store  
2                   Spa           Pizza Place             Restaurant  
3   Rental Car Location      Sculpture Garden          Boat or Ferry  
4     Indian Restaurant   Japanese Restaurant             Donut Shop

In [97]:
#start clustering the neighborhoods
# set number of clusters
kclusters = 6 #deciding to use 6 clusters

tor_grouped_clustering = tor_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(tor_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([4, 4, 4, 4, 0, 4, 4, 4, 4, 0], dtype=int32)

In [98]:
#add cluster number to the dataframe of neighborhoods
neigh_venues_sorted = neighborhoods_venues_sorted.drop('Cluster Labels', 1) 

# add clustering labels
neigh_venues_sorted.insert(1, 'Cluster Labels', kmeans.labels_)

tor_merged = toronto_data

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
tor_merged = tor_merged.join(neigh_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

tor_merged.head()
 

Postal Code   Latitude  Longitude          Borough  \
0         M4E  43.676357 -79.293031     East Toronto   
1         M4K  43.679557 -79.352188     East Toronto   
2         M4L  43.668999 -79.315572     East Toronto   
3         M4M  43.659526 -79.340923     East Toronto   
4         M4N  43.728020 -79.388790  Central Toronto   

                     Neighborhood  Cluster Labels 1st Most Common Venue  \
0                     The Beaches               0                 Trail   
1    The Danforth West, Riverdale               0      Greek Restaurant   
2  India Bazaar, The Beaches West               4                  Park   
3                 Studio District               4                  Café   
4                   Lawrence Park               5                  Park   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                   Pub     Health Food Store           Yoga Studio   
1    Italian Restaurant           Coffee Shop             Bookstore   
2  Fast Food Restaurant                   Pub          Liquor Store   
3           Coffee Shop             Gastropub                Bakery   
4           Swim School              Bus Line  Ethiopian Restaurant   

         5th Most Common Venue        6th Most Common Venue  \
0                Deli / Bodega  Eastern European Restaurant   
1               Ice Cream Shop                   Restaurant   
2               Sandwich Place                Burrito Place   
3                      Brewery          American Restaurant   
4  Eastern European Restaurant          Dumpling Restaurant   

     7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
0      Dumpling Restaurant            Donut Shop      Doner Restaurant   
1   Furniture / Home Store          Dessert Shop                   Pub   
2       Italian Restaurant            Restaurant        Ice Cream Shop   
3  Comfort Food Restaurant    Seafood Restaurant        Sandwich Place   
4               Donut Shop      Doner Restaurant               Dog Run   

  10th Most Common Venue  
0                Dog Run  
1   Caribbean Restaurant  
2             Steakhouse  
3            Cheese Shop  
4    Distribution Center

In [99]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(tor_merged['Latitude'], tor_merged['Longitude'], tor_merged['Neighborhood'], tor_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

**Map of the different clusters of neighborhoods in Toronto, clusters by types of venues they contain**

In [102]:
#cluster 1
tor_merged.loc[tor_merged['Cluster Labels'] == 0, tor_merged.columns[[4] + list(range(5, tor_merged.shape[1]))]]

Neighborhood  Cluster Labels  \
0                                         The Beaches               0   
1                        The Danforth West, Riverdale               0   
5                                    Davisville North               0   
6                   North Toronto West, Lawrence Park               0   
9   Summerhill West, Rathnelly, South Hill, Forest...               0   
13                          Regent Park, Harbourfront               0   
17                                 Central Bay Street               0   
35                             Parkdale, Roncesvalles               0   
37        Queen's Park, Ontario Provincial Government               0   

   1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
0                  Trail                   Pub     Health Food Store   
1       Greek Restaurant    Italian Restaurant           Coffee Shop   
5       Department Store                  Park           Pizza Place   
6            Coffee Shop        Clothing Store           Yoga Studio   
9                    Pub           Coffee Shop   Fried Chicken Joint   
13           Coffee Shop                  Park                Bakery   
17           Coffee Shop                  Café    Italian Restaurant   
35             Gift Shop        Breakfast Spot         Movie Theater   
37           Coffee Shop      Sushi Restaurant            Hobby Shop   

          4th Most Common Venue 5th Most Common Venue  \
0                   Yoga Studio         Deli / Bodega   
1                     Bookstore        Ice Cream Shop   
5                Breakfast Spot  Gym / Fitness Center   
6       Health & Beauty Service  Fast Food Restaurant   
9                   Pizza Place           Supermarket   
13                          Pub        Breakfast Spot   
17               Sandwich Place       Bubble Tea Shop   
35  Eastern European Restaurant               Dog Run   
37                Burrito Place                  Café   

          6th Most Common Venue   7th Most Common Venue 8th Most Common Venue  \
0   Eastern European Restaurant     Dumpling Restaurant            Donut Shop   
1                    Restaurant  Furniture / Home Store          Dessert Shop   
5                Sandwich Place                   Hotel     Food & Drink Shop   
6           Rental Car Location                    Café    Mexican Restaurant   
9              Sushi Restaurant                    Bank            Sports Bar   
13                         Café              Restaurant               Theater   
17                 Burger Joint                   Diner     Indian Restaurant   
35                          Bar      Italian Restaurant             Bookstore   
37          Distribution Center           Smoothie Shop    Italian Restaurant   

   9th Most Common Venue 10th Most Common Venue  
0       Doner Restaurant                Dog Run  
1                    Pub   Caribbean Restaurant  
5            Yoga Studio       Doner Restaurant  
6    Sporting Goods Shop     Chinese Restaurant  
9                   Café             Restaurant  
13          Dessert Shop             Shoe Store  
17   Japanese Restaurant             Donut Shop  
35            Restaurant           Dessert Shop  
37        Discount Store               Beer Bar

**Cluster 1 (cluster label 0) appears to consist mostly of coffee shops, clothing stores, and various types of restaurants**

In [103]:
#cluster 2
tor_merged.loc[tor_merged['Cluster Labels'] == 1, tor_merged.columns[[4] + list(range(5, tor_merged.shape[1]))]]

Neighborhood  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
22     Roselawn               1                Garden           Yoga Studio   

   3rd Most Common Venue 4th Most Common Venue        5th Most Common Venue  \
22      Department Store     Electronics Store  Eastern European Restaurant   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
22   Dumpling Restaurant            Donut Shop      Doner Restaurant   

   9th Most Common Venue 10th Most Common Venue  
22               Dog Run    Distribution Center

**Cluster 2 (cluster label 1) appears to be primarily gardens**

In [104]:
#cluster 3
tor_merged.loc[tor_merged['Cluster Labels'] == 2, tor_merged.columns[[4] + list(range(5, tor_merged.shape[1]))]]

Neighborhood  Cluster Labels  \
23  Forest Hill North & West, Forest Hill Road Park               2   

   1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
23                  Park         Jewelry Store                 Trail   

   4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
23      Sushi Restaurant              Bus Line           Yoga Studio   

   7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
23          Dessert Shop   Dumpling Restaurant            Donut Shop   

   10th Most Common Venue  
23       Doner Restaurant

**Cluster 3 (cluster label 2) is comprised of Forest Hill, which is a higher-end neighbourhood and as a result consists of parks, jewelry stores, sushi restaurants, etc.**

In [108]:
#cluster 4
tor_merged.loc[tor_merged['Cluster Labels'] == 3, tor_merged.columns[[4] + list(range(5, tor_merged.shape[1]))]]

Neighborhood  Cluster Labels 1st Most Common Venue  \
8   Moore Park, Summerhill East               3                  Park   
10                     Rosedale               3                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
8                  Trail                Lawyer            Restaurant   
10            Playground                 Trail          Dance Studio   

      5th Most Common Venue 6th Most Common Venue  \
8   Comfort Food Restaurant      Department Store   
10      Dumpling Restaurant            Donut Shop   

          7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
8   Eastern European Restaurant   Dumpling Restaurant            Donut Shop   
10             Doner Restaurant               Dog Run   Distribution Center   

   10th Most Common Venue  
8        Doner Restaurant  
10         Discount Store

**Cluster 4 (cluster label 3) appears to be primarily parks and trails**

In [107]:
#cluster 5
tor_merged.loc[tor_merged['Cluster Labels'] == 4, tor_merged.columns[[4] + list(range(5, tor_merged.shape[1]))]]

Neighborhood  Cluster Labels  \
2                      India Bazaar, The Beaches West               4   
3                                     Studio District               4   
7                                          Davisville               4   
11                        St. James Town, Cabbagetown               4   
12                               Church and Wellesley               4   
14                           Garden District, Ryerson               4   
15                                     St. James Town               4   
16                                        Berczy Park               4   
18                           Richmond, Adelaide, King               4   
19  Harbourfront East, Union Station, Toronto Islands               4   
20           Toronto Dominion Centre, Design Exchange               4   
21                     Commerce Court, Victoria Hotel               4   
24                The Annex, North Midtown, Yorkville               4   
25                     University of Toronto, Harbord               4   
26          Kensington Market, Chinatown, Grange Park               4   
27  CN Tower, King and Spadina, Railway Lands, Har...               4   
28                                     Stn A PO Boxes               4   
29             First Canadian Place, Underground city               4   
30                                           Christie               4   
31                       Dufferin, Dovercourt Village               4   
32                           Little Portugal, Trinity               4   
33       Brockton, Parkdale Village, Exhibition Place               4   
34                      High Park, The Junction South               4   
36                                 Runnymede, Swansea               4   
38  Business reply mail Processing Centre, South C...               4   

   1st Most Common Venue 2nd Most Common Venue          3rd Most Common Venue  \
2                   Park  Fast Food Restaurant                            Pub   
3                   Café           Coffee Shop                      Gastropub   
7         Sandwich Place          Dessert Shop                    Pizza Place   
11           Coffee Shop                  Café                         Bakery   
12           Coffee Shop            Restaurant                    Yoga Studio   
14        Clothing Store           Coffee Shop                           Café   
15                  Café             Gastropub                    Coffee Shop   
16           Coffee Shop          Cocktail Bar                    Cheese Shop   
18           Coffee Shop                  Café                   Concert Hall   
19           Coffee Shop              Aquarium                          Hotel   
20           Coffee Shop                  Café             Seafood Restaurant   
21                  Café           Coffee Shop                          Hotel   
24        Sandwich Place                  Café                    Coffee Shop   
25                  Café            Restaurant                            Bar   
26                  Café           Coffee Shop  Vegetarian / Vegan Restaurant   
27       Airport Service        Airport Lounge               Airport Terminal   
28           Coffee Shop                  Café                   Cocktail Bar   
29                  Café           Coffee Shop                     Restaurant   
30         Grocery Store                  Café                           Park   
31              Pharmacy                Bakery      Middle Eastern Restaurant   
32                   Bar      Asian Restaurant                     Restaurant   
33                  Café                Bakery                 Breakfast Spot   
34       Thai Restaurant    Mexican Restaurant                           Café   
36      Sushi Restaurant           Coffee Shop                    Pizza Place   
38    Light Rail Station           Yoga Studio           Gym / Fitness Center   

        4th Most Common Venue      

**Cluster 5 (cluster label 4) appears to consist mostly of food-based venues: cafes, coffee shops, and various types of restaurants**

In [109]:
#cluster 6
tor_merged.loc[tor_merged['Cluster Labels'] == 5, tor_merged.columns[[4] + list(range(5, tor_merged.shape[1]))]]

Neighborhood  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
4  Lawrence Park               5                  Park           Swim School   

  3rd Most Common Venue 4th Most Common Venue        5th Most Common Venue  \
4              Bus Line  Ethiopian Restaurant  Eastern European Restaurant   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
4   Dumpling Restaurant            Donut Shop      Doner Restaurant   

  9th Most Common Venue 10th Most Common Venue  
4               Dog Run    Distribution Center

**Cluster 6 (cluster label 5) consists of mostly a park, buses, and a swim school, and would appear to be a park-like area**